### ISSUE TITLE
DecRefError from mutable variable inside a closure

### ISSUE THREAD

**Issue by marcusroberts** [grain-lang/grain#1795]
The attached code crashes with a DecRefError when the parseString function is used in the match, even though the code is never run based on the input.

Commenting out line 20 or line 30 stops the error from occuring
(either       ```error = true``` or the match on '"' clause)

```
module Main

include "result"
include "string"
include "list"

let scanTokens = (source: String) => {
  let mut tokens = []
  let mut current = 0
  let mut start = 0
  let mut error = false

  let isAtEnd = () => current >= String.length(source)

  let parseString = () => {
    print("parsing a string")

    if (isAtEnd()) {
      print("Unterminated string")
      error = true
      return tokens
    }

    return tokens
  }

  let scanToken = () => {
    let c = ' '
    tokens = match (c) {
      '"' => parseString(),
      _ => {
        print("Unexpected character " ++ toString(c))
        error = true
        tokens
      },
    }
  }
  scanToken()
  tokens
}

scanTokens("{}")

```
(this is on main pre 0.6 release)

**Comment by ospencer** [comment]
Smaller repro:
```
module Main

let scanTokens = () => {
  let mut error = false

  let parseString = () => {
    error = true
  }

  let scanToken = () => {
    match (' ') {
      '"' => parseString(),
      _ => {
        error = true
      },
    }
  }
  scanToken()
}

scanTokens()

```

**Comment by ospencer** [comment]
I'm suspecting that the allocation_type is being too assertive with the box contained within the closure.

**Comment by phated** [comment]
I believe this was fixed by #1808 but it wasn't marked. Please reopen if I'm wrong.
